src/manifolds/Decorator.jl

```diff
@@ -0,0 +1,578 @@
+#
+# General functions for decorators
+#
+
+"""
+    AbstractDecoratorManifold <: Manifold
+"""
+abstract type AbstractDecoratorManifold <: Manifold end
+
+"""
+    is_default_decorator(M)
+
+For any manifold that is a subtype of [`AbstractDecoratorManifold`](@ref)
+indicates which function is used to determine the default decorator.
+This default decorator acts transparent even for functions where transparency is disabled.
+"""
+is_default_decorator(M::Manifold) = _is_default_decorator(M,val_is_default_decorator(M))
+_is_default_decorator(M::Manifold, ::Val{T}) where {T} = T
+val_is_default_decorator(M::Manifold) = Val(false)
+
+"""
+    is_decorator_transparent(M, f)
+
+Given a [`Manifold`](@ref) `M` and a function `f`, indicate, whether a
+[`AbstractDecoratorManifold`](@ref) acts transparent for `f`. This means, it
+just passes through down to the internally stored manifold.
+Only decorator manifolds can be transparent and their default is, to be transparent.
+A function that is affected by the decorator hence has to set this to `Val(false)`
+actively.
+If a decorator manifold is not in general transparent, it might still pass down
+for the case that a decorator is the default decorator, see [`is_default_decorator`](@ref).
+"""
+is_decorator_transparent(M::Manifold, f) = _is_decorator_transparent(M::Manifold, f, val_is_decorator_transparent(M,f))
+_is_decorator_transparent(M::Manifold, f, ::Val{T}) where {T} = T
+val_is_decorator_transparent(M::DT, f) where {DT <: Manifold} = Val(true)
+
+function manifold_function_not_implemented_message(M,f,x...)
+    s = join(map(string, map(typeof, x)),", "," and ")
```

Comment:
```suggestion
    s = join(map(string, map(typeof, x)), ", ", " and ")
```
It's slightly easier to ready with these spaces.